hphp/hack/src/hh_codesynthesis/agentGraphGenerator.py

```diff
@@ -0,0 +1,266 @@
+
+        similar_agents = 0
+        checked_agents = set()
+        similar_relations = []
+        for outer_agent, outer_dependents in enumerate(self.edges):
+            for inner_agent, inner_dependents in enumerate(
+                self.edges[outer_agent + 1 :]
+            ):
+                if (
+                    outer_dependents == inner_dependents
+                    and len(outer_dependents) != 0
+                    and inner_agent not in checked_agents
+                ):
+                    similar_agents += 1
+                    checked_agents.add(inner_agent)
+                    if inner_dependents not in similar_relations:
+                        similar_relations.append(inner_dependents)
+
+        print("There are {0} similar agents.".format(similar_agents))
+        print("Common dependents set is {0}".format(similar_relations))
+
+    def generate_raw(self, m: clingo.Model) -> None:
+        self._raw_model = m.__str__()
+
+    def on_model(self, m: clingo.Model) -> None:
+        raise RuntimeError("Must specify a valid method.")
+
+
+# Generate a set of agents using given distribution. The distribution has the
+# critical path requirement (levels) as well as how many roots are in the graph.
+def generating_agent_distribution(agent_distribution: List[int]) -> List[str]:
+    if not all(i > 0 for i in agent_distribution):
+        raise RuntimeError("Agent distribution must have all positive integers.")
+    # The levels in the generated agent graph is the length of this list.
+    return [
+        "agents({0}..{1}, {2}).".format(
+            sum(agent_distribution[:level]),
+            sum(agent_distribution[:level]) + number_of_agents_at_this_level - 1,
+            level,
+        )
+        for level, number_of_agents_at_this_level in enumerate(agent_distribution)
+    ]
+
+
+# Take in an agent distribution and a generator to create an agent graph.
+def generating_an_agent_graph(
+    agent_distribution: List[int], generator: AgentGraphGenerator
+) -> None:
+    # Logic programs for code synthesis.
+    asp_files = "hphp/hack/src/hh_codesynthesis"
+
+    # Clingo interfaces.
+    ctl = clingo.Control()
+
+    # Load LP for agent graph generating.
+    ctl.load(asp_files + "/agent_graph_generator.lp")
+    # Load LP for agent distribution.
+    ctl.add("base", [], "\n".join(generating_agent_distribution(agent_distribution)))
+    ctl.ground([("base", [])], context=AgentGraphClingoContext())
+
+    result: Union[clingo.solving.SolveHandle, clingo.solving.SolveResult] = ctl.solve(
+        on_model=generator.on_model
+    )
+    if isinstance(result, clingo.solving.SolveResult):
+        if result.unsatisfiable:
+            raise RuntimeError("Unsatisfiable.")
+
+
+def main() -> int:
+    agent_graph = AgentGraphGenerator()
+
+    agent_graph.on_model = agent_graph.generate_raw
+
+    # Parse the arguments
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--agents", nargs="+", type=int)
```

Comment:
Can we annotate this with a comment on what each number after the first number means?